<div class = "alert alert-info" >
<b>Приветствие</b>

Приветствую тебя, мой ревьюер! Я очень рада знакомству и совместной работе!

Меня зовут Дарья. Надеюсь, ты не против, если мы будем общаться на "ты"? Если тебя это смущает - дай знать!

Свои комментарии на замечания ревьюеров я, обычно, оставляю в синих блоках (таких, как этот). Если для тебя будет удобнее, чтобы я использовала другой цвет, то тоже - дай знать!

Буду очень признательна за комментарии к ошибкам, содержащие какое-то пояснение о том, как ошибки можно устранить.
    
Спасибо за твоё сотрудничество и комментарии - это очень ценно для меня!
    
Последнее задание на финальном спринте! Ура!

</div>


<div style="border:solid black 2px; padding: 20px">
<b>Дарья, привет!</b>

Меня зовут Полина Бережная, и я буду делать ревью твоего проекта. Давай будем общаться на «ты». 

Мне было очень приятно проверять твой проект, так как чувствуется твое старание и желание сделать задание на достойном уровне. Однако есть некоторые аспекты, которые требуют твоего внимания и доработки. 
    
В тетрадке ты можешь встретить 4 вида моих комментариев:
- <div class="alert alert-danger">
  
  **Комментарий ревьюера**
    
    Такой нюанс требует обязательной доработки, иначе проект не будет принят.
  </div>
    
    
- <div class="alert alert-warning">
  
    **Комментарий ревьюера**
    
    Так выделены не обязательные к доработке моменты. Mожешь учесть их при редактировании проекта и выполнении будущих заданий. На твое усмотрение.
  </div>
    
    
- <div class="alert alert-success">
    
  **Комментарий ревьюера** 
    
    Так отмечены удачные решения и подходы, на которые стоит опираться в будущих проектах.
  </div>
    
    
- <div style="border:solid black 2px; padding: 20px">
  
    **Чек-лист доработок**
    
    Такой комментарий будет находиться в конце тетрадки. В нем будут еще раз указаны нюансы, требующие обязательной корректировки.
  </div>

Предлагаю работать над проектом в форме диалога: если ты **что-то меняешь** в работе по моим рекомендациям — **пиши об этом** и помечай комментарий каким-то заметным цветом. Пожалуйста, **не перемещай, не изменяй и не удаляй мои комментарии**. Такой формат работы поможет выполнить повторную проверку твоего проекта оперативнее.
    
Не переживай, если что-то не получилось сделать идеально с первого раза. Основная цель нашей команды — не найти ошибки, а поделиться своим опытом и помочь тебе стать профи в аналитике 🙂
    
Жду твой проект на повторное ревью, ты уже почти на финише!

</div>

<div class = "alert alert-info" >
<b>Комментарий студента_v1:</b> Полина, очень рада знакомству! Спасибо большое за твоё ревью! Действительно, в одном месте я была невнимательна. Поправила твоё замечание. Комментарии выделены в таких же синих блоках.

</div>

# Проект по SQL

**Заказчик:**
Крупный сервис для чтения книг по подписке.

**Цель исследования:**

Проанализировать базу данных, содержащую информацию о книгах, издательствах, авторах, а также пользовательские обзоры книг для формулировки ценностного предложения для нового продукта.

**Исходные данные:**

* таблица с данными о книгах;
* таблица с данными об авторах;
* таблица с данными об издательствах;
* таблица с данными о пользовательских оценках книг;
* таблица с данными о пользовательских обзорах на книги.

**Задачи:**

* посчитать, сколько книг вышло после 1 января 2000 года;
* посчитать количество обзоров и среднюю оценку для каждой книги;
* определить издательство, которое выпустило наибольшее число книг толще 50 страниц (так мы исключим из анализа брошюры);
* определить автора с самой высокой средней оценкой книг (только для книг с 50 и более оценками);
* посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Оглавление <a id='Оглавление'></a>

### Подготовка данных к анализу
* [Загрузка библиотек](#1)
* [Подключение к базе](#2)
* [Загрузка и подготовка данных к анализу](#3)

### Исследование данных
* [Сколько книг вышло после 1 января 2000 года](#4)
* [Количество обзоров и средняя оценка для каждой книги](#5)
* [Определение издательства, которое выпустило наибольшее число книг толще 50 страниц](#6)
* [Определение автора с самой высокой средней оценкой книг (только для книг с 50 и более оценками)](#7)
* [Среднее количество обзоров от пользователей, которые поставили больше 50 оценок](#8)

### Общий вывод
* [Вывод из проведённого иследования](#9)

## Подготовка данных к анализу
<a id='1'></a>
### Загрузка библиотек 
[*Назад к оглавлению*](##Оглавление)

Импортируем все необходимые для решения поставленных задач библиотеки.

In [1]:
import pandas as pd 
from sqlalchemy import create_engine

<a id='2'></a>
### Подключение к базе
[*Назад к оглавлению*](##Оглавление)

Установим параметры подключения и сохраним коннектор.

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net', 
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], 
                                                         db_config['pwd'], 
                                                         db_config['host'], 
                                                         db_config['port'], 
                                                         db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

<a id='3'></a>
### Загрузка и подготовка данных к анализу
[*Назад к оглавлению*](##Оглавление)

Выполним sql-запрос ко всем таблицам и ознакомимся с данными.

In [3]:
def select(sql):
    return pd.io.sql.read_sql(sql, con = engine)

# таблица books
query = '''
SELECT * FROM books
'''
books = pd.io.sql.read_sql(query, con = engine)
display(books.head())
display(books.info())

# таблица authors
query_1 = '''
SELECT * FROM authors
'''
authors = pd.io.sql.read_sql(query_1, con = engine)
display(authors.head())
display(authors.info())

# таблица publishers
query_2 = '''
SELECT * FROM publishers
'''
publishers = pd.io.sql.read_sql(query_2, con = engine)
display(publishers.head())
display(publishers.info())

# таблица ratings
query_3 = '''
SELECT * FROM ratings
'''
ratings = pd.io.sql.read_sql(query_3, con = engine)
display(ratings.head())
display(ratings.info())

# таблица reviews
query_4 = '''
SELECT * FROM reviews
'''
reviews = pd.io.sql.read_sql(query_4, con = engine)
display(reviews.head())
display(reviews.info())

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

Таблица *books* содержит данные о книгах:
* `book_id` — идентификатор книги (тип данных - *int64*);
* `author_id` — идентификатор автора (тип данных - *int64*);
* `title` — название книги (тип данных - *object*);
* `num_pages` — количество страниц (тип данных - *int64*);
* `publication_date` — дата публикации книги (тип данных - *object*);
* `publisher_id` — идентификатор издателя (тип данных - *int64*).

Таблица *authors* содержит данные об авторах:
* `author_id` — идентификатор автора (тип данных - *int64*);
* `author` — имя автора (тип данных - *object*).

Таблица *publishers* содержит данные об издательствах:
* `publisher_id` — идентификатор издательства (тип данных - *int64*);
* `publisher` — название издательства (тип данных - *object*);

Таблица *ratings* содержит данные о пользовательских оценках книг:
* `rating_id` — идентификатор оценки (тип данных - *int64*);
* `book_id` — идентификатор книги (тип данных - *int64*);
* `username` — имя пользователя, оставившего оценку (тип данных - *object*);
* `rating` — оценка книги (тип данных - *int64*).

Таблица *reviews* содержит данные о пользовательских обзорах на книги:
* `review_id` — идентификатор обзора (тип данных - *int64*);
* `book_id` — идентификатор книги (тип данных - *int64*);
* `username` — имя пользователя, написавшего обзор (тип данных - *object*);
* `text` — текст обзора (тип данных - *object*).

<div class="alert alert-success">

**Комментарий ревьюера**
     
Здорово, что данные были изучены перед анализом 👍🏻

## Исследование данных
<a id='4'></a>
### Сколько книг вышло после 1 января 2000 года
[*Назад к оглавлению*](##Оглавление)

In [4]:
books_count = '''
SELECT COUNT(title) AS books_count FROM books
WHERE publication_date >= '2000-01-01'
'''
pd.io.sql.read_sql(books_count, con = engine)

books_count
0          821

После 1 января 2000 года было выпущено 821 книг.

<div class="alert alert-success">

**Комментарий ревьюера**
    
Запрос написан верно 👍🏻
    
P.S. Вообще авторы имели в виду период с 01.01.2000 включительно, но это не очевидно из формулировки задания. 

<div class = "alert alert-info" >
<b>Комментарий студента_v1:</b> Принято и поправлено. На две книги больше.

</div>

<a id='5'></a>
### Количество обзоров и средняя оценка для каждой книги
[*Назад к оглавлению*](##Оглавление)

In [5]:
ratings_reviews = '''
SELECT 
    books.title AS title,
    books.book_id AS id,
    AVG(ratings.rating) AS avg_rating,
    COUNT(DISTINCT reviews.text) AS cnt_reviews
FROM
    books
LEFT JOIN ratings ON ratings.book_id = books.book_id
LEFT JOIN reviews ON reviews.book_id = books.book_id
GROUP BY
     books.title,
     books.book_id
order by
     cnt_reviews DESC
'''
ratings_reviews = pd.io.sql.read_sql(ratings_reviews, con = engine)
ratings_reviews.head(10)

title   id  avg_rating  \
0                            Twilight (Twilight  #1)  948    3.662500   
1                                Water for Elephants  963    3.977273   
2                                   The Glass Castle  734    4.206897   
3  Harry Potter and the Prisoner of Azkaban (Harr...  302    4.414634   
4  The Curious Incident of the Dog in the Night-Time  695    4.081081   
5             The Da Vinci Code (Robert Langdon  #2)  696    3.830508   
6                                      The Alchemist  627    3.789474   
7                The Hobbit  or There and Back Again  750    4.125000   
8                                     The Book Thief  656    4.264151   
9  The Lightning Thief (Percy Jackson and the Oly...  779    4.080645   

   cnt_reviews  
0            7  
1            6  
2            6  
3            6  
4            6  
5            6  
6            6  
7            6  
8            6  
9            6

Наибольшее количество обзоров вышло по книге *"Twilight"* - 7. При этом рейтинг у этой книги ниже остальных книг с большим количеством обзоров: всего 3,66.

<div class="alert alert-danger">

**Комментарий ревьюера**
    
Все верно!
    
Но обрати внимание, что необходимо вывести количество обзоров и среднюю оценку для всех книг в базе. Результатирующая таблица не содержит нескольких произведений.
</div> 


<div class = "alert alert-info" >
<b>Комментарий студента_v1:</b> Согласна. Исправила. Теперь в таблице 1000 книг.

</div>

<a id='6'></a>
### Определение издательства, которое выпустило наибольшее число книг толще 50 страниц
[*Назад к оглавлению*](##Оглавление)

In [6]:
book_publisher = '''
SELECT
     publishers.publisher AS name_publisher,
     COUNT(books.book_id) AS count_of_published_books
FROM
    publishers
INNER JOIN books ON books.publisher_id = publishers.publisher_id
WHERE
    books.num_pages > 50
GROUP BY
    name_publisher
ORDER BY 
    count_of_published_books DESC
'''

book_publisher = pd.io.sql.read_sql(book_publisher, con = engine)
book_publisher.head()

name_publisher  count_of_published_books
0             Penguin Books                        42
1                   Vintage                        31
2  Grand Central Publishing                        25
3          Penguin Classics                        24
4          Ballantine Books                        19

Издательство Penguin Books выпустило наибольше число книг толще 50 страниц: 42 книги.

<div class="alert alert-success">

**Комментарий ревьюера**

Получен корректный ответ, молодец 👍🏻

<a id='7'></a>
### Определение автора с самой высокой средней оценкой книг (только для книг с 50 и более оценками)
[*Назад к оглавлению*](##Оглавление)

In [7]:
best_author = '''
SELECT
      SUBQ.author AS author,
      AVG(SUBQ.avg_rating) AS avg_rating
FROM
(SELECT
     authors.author AS author,
     books.book_id AS id,
     AVG(ratings.rating) AS avg_rating,
     COUNT(ratings.rating) AS cnt_rating
FROM
    authors
INNER JOIN books ON books.author_id = authors.author_id
INNER JOIN ratings ON ratings.book_id = books.book_id
GROUP BY
    author,
    id
    HAVING
    COUNT(ratings.rating) >= 50) AS SUBQ
GROUP BY
    author
ORDER BY
    avg_rating DESC
'''

best_author = pd.io.sql.read_sql(best_author, con = engine)
best_author

author  avg_rating
0                          J.K. Rowling/Mary GrandPré    4.283844
1                   Markus Zusak/Cao Xuân Việt Khương    4.264151
2                                      J.R.R. Tolkien    4.258446
3                                   Louisa May Alcott    4.192308
4                                        Rick Riordan    4.080645
5                                     William Golding    3.901408
6                                       J.D. Salinger    3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...    3.787879
9                                           Dan Brown    3.754540
10                                         Lois Lowry    3.750000
11            George Orwell/Boris Grabnar/Peter Škerl    3.729730
12                                    Stephenie Meyer    3.662500
13                                     John Steinbeck    3.622951

Cамая высокая средняя оценка у автора J. K. Rowling / Mary GrandPré - 4,28.

<div class="alert alert-success">

**Комментарий ревьюера**

Абсолютно верно 👍🏻

<a id='8'></a>
### Среднее количество обзоров от пользователей, которые поставили больше 50 оценок
[*Назад к оглавлению*](##Оглавление)

In [8]:
avg_rating_review = '''
SELECT ROUND(avg(count)) 
FROM
(SELECT
    COUNT(text)
FROM 
    reviews
FULL OUTER JOIN (SELECT username,
                COUNT(rating_id) as cnt
FROM 
    ratings
GROUP BY
        username) AS cnt_rating
ON cnt_rating.username = reviews.username
where cnt > 50
GROUP BY reviews.username) as count;                                  
'''

pd.io.sql.read_sql(avg_rating_review, con = engine)

round
0   24.0

Среднее количество обзоров пользователей, которые поставили более чем по 50 оценок, равно 24.

<div class="alert alert-success">

**Комментарий ревьюера**
     
Отлично 🔥

## Общий вывод
<a id='9'></a>

<div style="border:solid Red 2px; padding: 40px">
<h3> Вывод <a class="tocSkip"> </h3>
    
<br/>
Для формулировки ценностного предложения для нового продукта необходимо принять во внимание следующую информацию:
<li>в таблице <b>book_publisher</b> лежит список всех издательств, которые выпускают книги от 50 страниц;</li>
<li>таблица <b>best_author</b> содержит список авторов с самой выской средней оценкой, при этом у этих авторов учтены книги которые имеют больше 50 пользовательских оценок;</li>
<li>в таблице <b>ratings_reviews</b>, есть книги со средней оценкой и пользовательскими обзорами: стоит обратить внимание на книги, где кол-во обзоров больше 3, так как достаточно много завышенных оценок с 2 обзорами;</li>
<li>среднее количество текстовых обзоров на книги более чем 50 оценками имеют в среднем 24 текстовых обзора (что будет достаточно информативно для будущих пользователей).</li>

<br/>

<div style="border:solid black 2px; padding: 20px">
<h3>Чек-лист доработок </h3> 
    
По проекту видно, что в него было вложено много времени и сил! 
 
**Однако нам нужно проработать запрос 2.**
    
Жду твой проект на повторное ревью, осталось совсем чуть-чуть. Не стесняйся оставлять комментарии и задавать вопросы. Удачи! Все получится 🙂

<div class = "alert alert-info" >
<b>Комментарий студента_v1:</b> Ещё раз благодарю тебя за помощь и советы! Жду твоего вердикта)))

</div>